services/api/src/utils/search.js

```diff
@@ -0,0 +1,102 @@
+const Joi = require('@hapi/joi');
+const ObjectId = require('mongoose').Types.ObjectId;
+const PassThrough = require('stream').PassThrough;
+const csv = require('fast-csv');
+
+function searchValidation(options = {}) {
+  const { limit = 50, sortField = 'createdAt', sortOrder = 'desc' } = options;
+  return {
+    startAt: Joi.date(),
+    endAt: Joi.date(),
+    skip: Joi.number().default(0),
+    sort: Joi.object({
+      field: Joi.string().required(),
+      order: Joi.string().required(),
+    }).default({
+      field: sortField,
+      order: sortOrder,
+    }),
+    limit: Joi.number().positive().default(limit),
+    keywords: Joi.string(),
+  };
+}
+
+function exportValidation(options = {}) {
+  const { filename = 'export.csv' } = options;
+  return {
+    filename: Joi.string().default(filename),
+    format: Joi.string().allow('json', 'csv').default('json'),
+  };
+}
+
+function getSearchQuery(body, options = {}) {
+  const { keywords, startAt, endAt } = body;
+  const query = { deletedAt: { $exists: false } };
+  if (startAt || endAt) {
+    query.createdAt = {};
+    if (startAt) {
+      query.createdAt.$gte = startAt;
+    }
+    if (endAt) {
+      query.createdAt.$lte = endAt;
+    }
+  }
+  if (keywords && options.keywordsFields) {
```

Comment:
perhaps add a warning if options.keywordsFields is not set but a keywords is ?

Reply:
Added!